### PR TITLE
feat(ingestion): Adds more advanced configurations for runtime debugging

### PR DESCRIPTION
**Pull request by pedro93**
## Checklist

Adds advanced configurations for specifying environment variables, DataHub plugins and pip packages at runtime for managed ingestion.

This feature is specific to managed ingestion as it uses logic in datahub actions.
Tested manually.
Here is a screenshot of how it looks:

Screenshot 2023-10-11 at 23 16 36

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [x] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [x] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: docker/build.gradle [hunk 1]
```diff
@@ -97,10 +97,20 @@ task quickstartDebug(type: Exec, dependsOn: ':metadata-ingestion:install') {
...
+    // Elastic
+    // environment "DATAHUB_SEARCH_IMAGE", 'elasticsearch'
+    // environment "DATAHUB_SEARCH_TAG", '7.10.1'
+
+    // OpenSearch
+    environment "DATAHUB_SEARCH_IMAGE", 'opensearchproject/opensearch'
+    environment "DATAHUB_SEARCH_TAG", '2.9.0'
+    environment "XPACK_SECURITY_ENABLED", 'plugins.security.disabled=true'
+    environment "USE_AWS_ELASTICSEARCH", 'true'
+
```
Comment (chriscollins3456) [inline docker/build.gradle:112]: why are these changes necessary?

File: datahub-web-react/src/app/ingest/source/IngestionSourceList.tsx [hunk 2]
```diff
@@ -173,14 +173,33 @@ export const IngestionSourceList = () => {
...
+                    config: {
+                        ...input.config,
+                        // This is needed to ensure that graphQL __typename properties are not passed to the
+                        // updateIngestionSource GraphQL call which would fail otherwise as it is not an expected
+                        // input type.
+                        extraArgs: sanitizeExtraArgs(input.config?.extraArgs),
+                    },
+                };
+            }
+            console.log(`[updateIngestionSource] ${JSON.stringify(sanitizedInput)}`);
```
Comment (chriscollins3456) [inline datahub-web-react/src/app/ingest/source/IngestionSourceList.tsx:201]: we try to avoid consoling stuff like this out so I'd just remove

File: datahub-web-react/src/app/ingest/source/IngestionSourceList.tsx [hunk 3]
```diff
@@ -173,14 +173,33 @@ export const IngestionSourceList = () => {
         setFocusSourceUrn(undefined);
     };
 
+    const sanitizeExtraArgs = (extraArgs): StringMapEntryInput[] => {
```
Comment (chriscollins3456) [inline datahub-web-react/src/app/ingest/source/IngestionSourceList.tsx:176]: maybe `formatExtraArgs` instead of "sanitize" - sanitize implies cleaning for security purposes like when we sanitize rich text before rendering it on the UI

File: datahub-web-react/src/app/ingest/source/IngestionSourceList.tsx [hunk 4]
```diff
@@ -294,6 +313,7 @@ export const IngestionSourceList = () => {
                             (recipeBuilderState.config?.executorId as string)) ||
                         DEFAULT_EXECUTOR_ID,
                     debugMode: recipeBuilderState.config?.debugMode || false,
+                    extraArgs: recipeBuilderState.config?.extraArgs || [],
```
Comment (chriscollins3456) [inline datahub-web-react/src/app/ingest/source/IngestionSourceList.tsx:316]: can you do the formatting of `extraArgs` here and pass into `createOrUpdateIngestionSource` properly formatted instead of having to do the whole shebang where you have to spread the whole object etc.?

so you could just do `extraArgs: sanitizeExtraArgs(recipeBuilderState.config?.extraArgs || []),`

File: datahub-web-react/src/app/ingest/source/builder/NameSourceStep.tsx [hunk 5]
```diff
@@ -116,6 +204,39 @@ export const NameSourceStep = ({ state, updateState, prev, submit }: StepProps)
...
+                            <Typography.Paragraph>
+                                Advanced: Add extra pip libraries for an ingestion execution
+                            </Typography.Paragraph>
+                            <Input
+                                data-testid="extra-pip-reqs-input"
+                                placeholder='["sqlparse==0.4.3"]'
+                                value={retrieveExtraReqs()}
+                                onChange={(event) => setExtraReqs(event.target.value)}
+                            />
+                        </Form.Item>
```
Comment (chriscollins3456) [inline datahub-web-react/src/app/ingest/source/builder/NameSourceStep.tsx:239]: so im trying to get better at enforcing small component because that keeps the frontend so much cleaner and easier to reason about. I'd move each of these inputs into their own new small file and component where in each of those the getter and setter logic can live. ie. `retrieveExtraReqs` and `setExtraReqs` can live in its own component where you render the input for Extra Pip Libraries

File: metadata-models/src/main/pegasus/com/linkedin/ingestion/DataHubIngestionSourceInfo.pdl [hunk 6]
```diff
@@ -56,5 +56,10 @@ record DataHubIngestionSourceInfo {
       * Whether or not to run this ingestion source in debug mode
       */
       debugMode: optional boolean
+
+     /**
+      * Extra arguments for the ingestion run.
+      */
+      extraArgs: optional map[string, string]
```
Comment (chriscollins3456) [inline metadata-models/src/main/pegasus/com/linkedin/ingestion/DataHubIngestionSourceInfo.pdl:63]: did you have a specific reason you wanted this to be a loose map instead of several stronger typed fields for the new things you want to add?

**Review by chriscollins3456 — COMMENTED** [review]
looking good! I have some suggestions but nothing major on my end

File: metadata-models/src/main/pegasus/com/linkedin/ingestion/DataHubIngestionSourceInfo.pdl (hunk 6 above)
Comment (pedro93) [reply metadata-models/src/main/pegasus/com/linkedin/ingestion/DataHubIngestionSourceInfo.pdl:63]: It maps directly to the underlying PDL model, I didn't want to be prescriptive here. Do you have a strong opinion on this?

File: docker/build.gradle (hunk 1 above)
Comment (pedro93) [reply docker/build.gradle:112]: This is to make quickstartDebug gradlew task work. I just made it consistent with `./gradlew quickstart`

File: metadata-models/src/main/pegasus/com/linkedin/ingestion/DataHubIngestionSourceInfo.pdl (hunk 6 above)
Comment (chriscollins3456) [reply metadata-models/src/main/pegasus/com/linkedin/ingestion/DataHubIngestionSourceInfo.pdl:63]: nope that's cool with me

**Review by chriscollins3456 — APPROVED** [review]
cool beans
